src/pal/src/thread/thread.cpp

```diff
@@ -1353,82 +1384,177 @@ GetThreadTimes(
         &pobjThread
         );
     
-	if (palError != NO_ERROR)
+    if (palError != NO_ERROR)
     {
         ASSERT("Unable to get thread data from handle %p"
               "thread\n", hThread);
         SetLastError(ERROR_INTERNAL_ERROR);
-        goto GetThreadTimesExit;
-	}	
+        goto SetTimesToZero;
+    }   
 
     pthrTarget->Lock(pthrCurrent);
-	
+    
     mach_port_t mhThread;
     mhThread = pthread_mach_thread_np(pthrTarget->GetPThreadSelf());
-	
-	kern_return_t status;
-	status = thread_info(
-	    mhThread, 
-		THREAD_BASIC_INFO, 
-		(thread_info_t)&resUsage, 
-		&resUsage_count);
-	
+    
+    kern_return_t status;
+    status = thread_info(
+        mhThread, 
+        THREAD_BASIC_INFO, 
+        (thread_info_t)&resUsage, 
+        &resUsage_count);
+
+    pthrTarget->Unlock(pthrCurrent);
+    
     if (status != KERN_SUCCESS)
-	{
+    {
         ASSERT("Unable to get resource usage information for the current "
               "thread\n");
         SetLastError(ERROR_INTERNAL_ERROR);
-        goto GetThreadTimesExit;
+        goto SetTimesToZero;
     }
+
+    /* Get the time of user mode execution, in nanoseconds */
+    calcTime = (__int64)resUsage.user_time.seconds * SECS_TO_NS;
+    calcTime += (__int64)resUsage.user_time.microseconds * USECS_TO_NS;
+    /* Assign the time into lpUserTime */
+    lpUserTime->dwLowDateTime = (DWORD)calcTime;
+    lpUserTime->dwHighDateTime = (DWORD)(calcTime >> 32);
+
+    /* Get the time of kernel mode execution, in nanoseconds */
+    calcTime = (__int64)resUsage.system_time.seconds * SECS_TO_NS;
+    calcTime += (__int64)resUsage.system_time.microseconds * USECS_TO_NS;
+    /* Assign the time into lpKernelTime */
+    lpKernelTime->dwLowDateTime = (DWORD)calcTime;
+    lpKernelTime->dwHighDateTime = (DWORD)(calcTime >> 32);
+
+    retval = TRUE;
+
+    goto GetThreadTimesInternalExit;
+
+#else //HAVE_MACH_THREADS
+
+    if (!IsCurrentThread(hThread))
+    {
+        //Set all times to 0 if this isn't current thread
+        goto SetTimesToZero;
+    }
+
+#if HAVE_CLOCK_THREAD_CPUTIME
+    struct timespec ts;
+    if (clock_gettime(CLOCK_THREAD_CPUTIME_ID, &ts) != 0)
+    {
+        ASSERT("clock_gettime(CLOCK_THREAD_CPUTIME_ID) failed; errno is %d (%s)\n", errno, strerror(errno));
+        goto SetTimesToZero;
+    }
+
+    if (lpUserTime)
```

Comment:
Since GetThreadTimesInternal is an internal function, we don't need this check. We should assert that it is not NULL instead. The caller must pass a valid parameter.